src/moi.jl

```diff
@@ -406,13 +413,186 @@ function build_bilevel(
+                        bilinear_upper_dual_to_lower_primal,
+                        V,
+                        w,
+                        bilinear_upper_dual_to_quad_term,
+                        lower_obj_terms,
+                        lower_to_m_idxmap,
+                        lower_primal_dual_map,
+                        lower_dual_idxmap
+                    )
+                    if isnothing(linearizations)
+                        @warn("Unable to linearize bilinear terms due to underdertermined system of equations. Skipping linearization process.")
+                    end
+                else
+                    @warn("Required conditions for linearization not met. Skipping linearization process.")
+                end
+            else  # AB_N is not empty
+                @debug("set AB_N is NOT empty")
+                
+                # TODO input flag for checking for blocks? (to save time)
+                num_blocks, rows, cols = find_blocks(V, U)
+
+                conditions_passed = Bool[]
+                nrows, ncols = size(V)
+                for n in 1:num_blocks
+                    # TODO can skip blocks that are not linked to bilinear terms?
+                    Vblock = spzeros(nrows, ncols)
+                    Vblock[rows[n],:] = V[rows[n],:]
+                    check = check_non_empty_AB_N_conditions(
+                        intersect(J_U, rows[n]), U, intersect(N_U, cols[n]), intersect(A_N, cols[n]), B, Vblock, 
+                        lower_primal_var_to_lower_con, upper_var_to_lower_ctr,
+                        bilinear_upper_dual_to_quad_term, 
+                        bilinear_upper_dual_to_lower_primal)
+                    push!(conditions_passed, check)
+                end
+                # recover memory
+                rows, cols = nothing, nothing
+                @debug("Done looping over V blocks")
+                
+                if all(conditions_passed)
+                    linearizations = linear_terms_for_non_empty_AB(
+                        lower,
+                        upper_var_to_lower_ctr,
+                        bilinear_upper_dual_to_lower_primal,
+                        V,
+                        w,
+                        A_N,
+                        bilinear_upper_dual_to_quad_term,
+                        lower_obj_terms,
+                        lower_to_m_idxmap,
+                        lower_primal_dual_map,
+                        lower_dual_idxmap
+                    )
+                    if isnothing(linearizations)
+                        @warn("Unable to linearize bilinear terms due to underdertermined system of equations. Skipping linearization process.")
+                    end
+                else
+                    @warn("Required conditions for linearization not met. Skipping linearization process.")
+                end
+
+            end
+
+            # LINEARIZATION
+            # TODO switch signs with MAX sense?
```

Comment:
maybe for now add an assert that we have a min sense

Reply:
done in 3f96626ceef9713d84f5585e3a2d9621d843d8bb